metaflow/datatools/s3.py

```diff
@@ -382,100 +467,261 @@ def get(self, key=None, return_missing=False):
+                            return it as an S3Object with .exists == False.
+            return_info: (optional, default False) if set to True, fetch the
+                         content-type and user metadata associated with the object.
+
+        Returns:
+            an S3Object corresponding to the object requested.
+        """
+        url, r = self._url_and_range(key)
+        src = urlparse(url)
+
+        def _download(s3, tmp):
+            if r:
+                resp = s3.get_object(
+                    Bucket=src.netloc,
+                    Key=src.path.lstrip('/'),
+                    Range=r)
+                code = str(resp['ResponseMetadata']['HTTPStatusCode'])
+                if code[0] == '2':
+                    with open(tmp, mode='w') as t:
+                        t.write(resp['Body'].read())
+                else:
+                    # TODO: Better raised error
+                    raise RuntimeError("Could not load file")
+            else:
+                s3.download_file(src.netloc, src.path.lstrip('/'), tmp)
+                return url
+
+        def _info(s3, tmp):
+            resp = s3.head_object(Bucket=src.netloc, Key=src.path.lstrip('/"'))
+            with open('%s' % tmp, mode='w') as f:
+                args = {
+                    'content_type': resp['ContentType'],
+                    'metadata': resp['Metadata']}
+                json.dump(args, f)
+
+        path_info = None
         try:
             path = self._one_boto_op(_download, url)
+            if return_info:
+                path_info = self._one_boto_op(_info, url)
         except MetaflowS3NotFound:
             if return_missing:
                 path = None
             else:
                 raise
-
+        if path_info:
+            with open(path_info, 'r') as f:
+                info = json.load(f)
+            return S3Object(
+                self._s3root, url, path,
+                content_type=info['content_type'],
+                metadata=info['metadata'])
         return S3Object(self._s3root, url, path)
 
-    def get_many(self, keys, return_missing=False):
+    def get_many(self, keys, return_missing=False, return_info=False):
         """
         Get many objects from S3 in parallel.
 
         Args:
-            keys: (required) a list of suffixes identifying the objects.
+            keys: (required) a list of suffixes identifying the objects. Each
+                  item in the list can also be an object containing the properties
+                  `key`, `offset` and `length to specify a range query.
+                  `S3GetObject` is such an object.
             return_missing: (optional, default False) if set to True, do
                             not raise an exception for a missing key but
                             return it as an S3Object with .exists == False.
+            return_info: (optional, default False) if set to True, fetch the
```

Comment:
don't we get `content-type` and headers (i.e. user metadata) for free on the `s3op` side? If so, can't we just return info always by default?

Reply:
I was trying to be consistent between the calls (get vs get_many). It is also not *entirely* free (close to free but not totally) due to the additional file mumbo jumbo to get the data back out. I am fine making it true by default but would prefer to keep consistency (so making it true by default for the get too which does imply an additional call).

Reply:
Done